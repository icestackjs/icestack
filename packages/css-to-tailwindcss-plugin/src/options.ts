import path from 'node:path'
import { pathToFileURL } from 'node:url'
import defu from 'defu'
import type { IProcessOptions } from './types'

export function getDefaults(): Required<IProcessOptions> {
  const pkgPath = require.resolve('tailwindcss/package.json')
  const nodeModulesPath = path.dirname(path.dirname(pkgPath))
  return {
    atImportOptions: {},
    sassOptions: {
      importers: [
        {
          // https://github.com/webpack-contrib/sass-loader/blob/master/src/index.js#L55
          findFileUrl(url) {
            const twUrl = pathToFileURL(path.resolve(nodeModulesPath, url))
            return new URL(twUrl)
          }
        }
      ]
    },
    tailwindcssConfig: '',
    generatorOptions: {}
  }
}

export function getOptions(opts?: IProcessOptions): Required<IProcessOptions> {
  return defu(opts, getDefaults())
}
