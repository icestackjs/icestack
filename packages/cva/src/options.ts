import defu from 'defu'
import type { InternalOptions, UserDefinedOptions } from './types'

export function getOptions(opts: UserDefinedOptions = {}) {
  const options: InternalOptions = defu<InternalOptions, InternalOptions[]>(opts, {
    prefix: '',
  })
  return options
}
