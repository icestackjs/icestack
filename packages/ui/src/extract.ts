import path from 'node:path'
import plugin from 'tailwindcss/plugin'
import { set } from 'lodash'
import { CssInJs } from 'postcss-js'
import merge from 'merge'
import objHash from 'object-hash'
import { getColors } from './colors'
import { walkScssSync } from './utils'
import { extractScss } from '@/sass'
import { resolveJsDir, scssDir } from '@/dirs'
import { someExtends } from '@/constants'
import { CodegenOptions } from '@/types'
import { getFileCache } from '@/cache'

export type IOptions = {
  options: CodegenOptions
  outSideLayerCss: 'base' | 'utilities' | 'components'
}

function groupedComs(
  resultArray: {
    key: string
    value: CssInJs
  }[]
) {
  return resultArray.reduce<Record<string, Record<string, CssInJs>>>((acc, cur) => {
    const [type, key] = cur.key.split('/')

    // style / unstyle / global
    if (acc[key]) {
      acc[key][type] = cur.value
    } else {
      acc[key] = {
        [type]: cur.value
      }
    }

    return acc
  }, {})
}

export function getJsObj(opts: IOptions) {
  const { outSideLayerCss, options } = opts // defu<IOptions, Partial<IOptions>[]>(opts, {})
  const { outdir, base } = options
  // await ensureDir(pluginsDir)
  switch (outSideLayerCss) {
    case 'base': {
      const cache = getFileCache('base/index')
      const objhash = objHash(base)
      if (cache.getKey('objhash') === objhash) {
        return cache.getKey('value')
      } else {
        const basePath = path.resolve(scssDir, 'base')
        const resultArray: { key: string; value: CssInJs }[] = []
        for (const file of walkScssSync(basePath)) {
          resultArray.push({
            key: path.relative(basePath, file.path).replace(/\.scss$/, ''),
            value: extractScss({
              outdir,
              filename: file.path,
              outSideLayerCss,
              options
            })
          })
        }
        const value = merge.recursive(true, ...resultArray.map((x) => x.value))
        cache.setKey('value', value)
        cache.setKey('objhash', objhash)
        cache.save()
        return value
      }
    }
    case 'utilities': {
      const utilitiesPath = path.resolve(scssDir, 'utilities')
      const resultArray: { key: string; value: CssInJs }[] = []

      for (const file of walkScssSync(path.resolve(utilitiesPath, 'global'))) {
        resultArray.push({
          key: path.relative(utilitiesPath, file.path).replace(/\.scss$/, ''),
          value: extractScss({
            outdir,
            filename: file.path,
            outSideLayerCss,
            resolveConfig(config) {
              set(config, 'theme.extend.colors', getColors(options))
            },
            options
          })
        })
      }

      const utilitiesJs = path.resolve(resolveJsDir(outdir), 'utilities')

      for (const file of walkScssSync(utilitiesPath)) {
        resultArray.push({
          key: path.relative(utilitiesPath, file.path).replace(/\.scss$/, ''),
          value: extractScss({
            outdir,
            filename: file.path,
            outSideLayerCss,
            resolveConfig(config) {
              set(config, 'theme.extend.colors', getColors(options))
              config.plugins = [
                plugin(
                  ({ addUtilities }) => {
                    const obj = require(path.resolve(utilitiesJs, 'index.js'))
                    // @ts-ignore
                    addUtilities(Object.values(obj))
                  },
                  {
                    theme: {
                      extend: {
                        ...someExtends
                      }
                    }
                  }
                )
              ]
            },
            options
          })
        })
      }

      return groupedComs(resultArray)
    }
    case 'components': {
      const utilitiesJs = path.resolve(resolveJsDir(outdir), 'utilities')

      const resultArray: { key: string; value: CssInJs }[] = []
      const fromDir = path.resolve(scssDir, 'components')
      for (const file of walkScssSync(fromDir)) {
        resultArray.push({
          key: path.relative(fromDir, file.path).replace(/\.scss$/, ''),
          value: extractScss({
            outdir,
            filename: file.path,
            resolveConfig: (config) => {
              set(config, 'theme.extend.colors', getColors(options))
              config.plugins = [
                plugin(
                  ({ addUtilities }) => {
                    const obj = require(path.resolve(utilitiesJs, 'index.js'))
                    // @ts-ignore
                    addUtilities(Object.values(obj))
                  },
                  {
                    theme: {
                      extend: {
                        ...someExtends
                      }
                    }
                  }
                )
              ]
            },
            outSideLayerCss,
            options
          })
        })
      }

      return groupedComs(resultArray)
    }
    default:
  }
}

export function extractAll(options: CodegenOptions) {
  const base = getJsObj({
    options,
    outSideLayerCss: 'base'
  })
  console.log('extract base finished!')
  const utilities = getJsObj({
    options,
    outSideLayerCss: 'utilities'
  })
  console.log('extract utilities finished!')
  const components = getJsObj({
    options,
    outSideLayerCss: 'components'
  })
  console.log('extract components finished!')
  return {
    base,
    utilities,
    components
  }
}
