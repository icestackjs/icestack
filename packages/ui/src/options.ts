import defu from 'defu'
import { getCodegenDefaults } from './defaults'
import { CodegenOptions, DeepPartial } from './types'

export function getCodegenOptions(options?: DeepPartial<CodegenOptions>, raw?: boolean): CodegenOptions {
  let presets: DeepPartial<CodegenOptions>[] = []
  if (options?.presets && Array.isArray(options?.presets)) {
    // @ts-ignore
    presets = options.presets ?? []
  }
  return defu<CodegenOptions, DeepPartial<CodegenOptions>[]>(options, ...presets, getCodegenDefaults(raw))
}
