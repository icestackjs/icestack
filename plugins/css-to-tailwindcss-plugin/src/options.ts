import type { IProcessOptions, TailwindcssPluginOptions } from './types'
import path from 'node:path'
import process from 'node:process'
import { pathToFileURL } from 'node:url'
import defu from 'defu'

function getDefaultCacheDir() {
  return path.resolve(process.cwd(), 'node_modules', '.css-to-tailwindcss-plugin')
}

export function getDefaults(): IProcessOptions & { cacheDir?: string } {
  const pkgPath = require.resolve('tailwindcss/package.json')
  const nodeModulesPath = path.dirname(path.dirname(pkgPath))
  return {
    atImportOptions: {},
    sassOptions: {
      importers: [
        {
          // https://github.com/webpack-contrib/sass-loader/blob/master/src/index.js#L55
          findFileUrl(url) {
            const twUrl = pathToFileURL(path.resolve(nodeModulesPath, url))
            return new URL(twUrl)
          },
        },
      ],
    },
    tailwindcssResolved: false,
    withOptions: true,
    withOptionsWalkCSSRuleObject(obj) {
      return obj
    },
    cacheDir: getDefaultCacheDir(),
  }
}

export function getOptions(opts?: IProcessOptions) {
  return defu(opts, getDefaults()) as Required<TailwindcssPluginOptions>
}
