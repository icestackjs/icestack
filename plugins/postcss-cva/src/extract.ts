import type { PluginCreator } from 'postcss'
import { get, set } from 'lodash'
import { defu, objHash } from '@icestack/shared'
import type { CvaParams, CvaParamsSet } from './types'
import { cvaSymbol, defaultParser, extractParams, getParentRule, pickComment, setAdd } from './regex'

export interface ExtractOption {
  prefix: string
  process: (res?: CvaParams) => void
  remove: boolean
  filter: (id: unknown) => boolean
}

const creator: PluginCreator<Partial<ExtractOption>> = (opts) => {
  const {
    process,
    prefix: _prefix,
    remove,
    filter,
  } = defu<ExtractOption, Partial<ExtractOption>[]>(opts, {
    filter: () => true,
    remove: true,
  })

  return {
    postcssPlugin: 'postcss-icestack-extract-cva-params-plugin',
    prepare(res) {
      const result: CvaParamsSet = {
        base: new Set<string>(),
        variants: {},
        compoundVariants: [],
        defaultVariants: {},
        meta: {},
      }
      const prefix = _prefix ?? ''
      const hashMap = new Map<string, CvaParamsSet['compoundVariants'][number]>()
      const weakMap = new WeakMap()

      function addBase(value: string | string[]) {
        setAdd(result.base, value)
      }

      function addVariant(
        value: string | string[],
        entries: [
          string,
          {
            value: string
          },
        ][],
      ) {
        for (const [p1, { value: p2 }] of entries) {
          const p = `${p1}.${p2}`
          const arr = get(result.variants, p)

          if (arr instanceof Set) {
            setAdd(arr, value)
          }
          else {
            const st = new Set<string>()
            setAdd(st, value)
            set(result.variants, p, st)
          }
        }
      }

      function addDefaultVariant(
        entries: [
          string,
          {
            value: string
          },
        ][],
      ) {
        for (const [key, { value }] of entries) {
          set(result.defaultVariants, key, value)
        }
      }

      function addCompoundVariant(
        value: string | string[],
        entries: [
          string,
          {
            value: string
          },
        ][],
        hashCode: string,
      ) {
        const item = hashMap.get(hashCode)
        if (item) {
          item.class && setAdd(item.class, value)
        }
        else {
          const set = new Set<string>()
          setAdd(set, value)

          hashMap.set(
            hashCode,
            // @ts-ignore
            entries.reduce(
              (acc, [k, { value }]) => {
                // @ts-ignore
                acc[k] = value
                return acc
              },
              {
                class: set,
              },
            ),
          )
        }
      }

      const filename = res.root.source?.input.file
      if (filename) {
        const skip = !filter(filename)
        if (skip) {
          return {}
        }
      }

      // res.root
      return {
        Comment(comment) {
          const res = pickComment(comment)
          if (res) {
            weakMap.set(comment, cvaSymbol)
            const { next, suffix, type } = res
            const { query, params } = extractParams(suffix, res)
            const hashCode = objHash(query)
            const entries = Object.entries(query)

            if (next) {
              const rule = getParentRule(comment)

              if (rule) {
                const ast = defaultParser.astSync(rule.selector)

                let value: string | undefined
                ast.walkClasses((cls) => {
                  value = cls.value
                  // return false
                })

                if (value) {
                  // value = prefix + value
                  switch (type) {
                    case 'base': {
                      addBase(value)

                      break
                    }
                    case 'variant': {
                      addVariant(value, entries)

                      break
                    }
                    case 'compoundVariant': {
                      addCompoundVariant(value, entries, hashCode)

                      break
                    }
                  }
                }
              }
            }
            else {
              switch (type) {
                case 'defaultVariant': {
                  addDefaultVariant(entries)

                  break
                }
                case 'base': {
                  addBase(params.map(x => prefix + x))

                  break
                }
                case 'variant': {
                  addVariant(
                    params.map(x => prefix + x),
                    entries,
                  )

                  break
                }
                case 'compoundVariant': {
                  addCompoundVariant(
                    params.map(x => prefix + x),
                    entries,
                    hashCode,
                  )

                  break
                }
                case 'meta': {
                  for (const [key, { value }] of entries) {
                    result.meta[key] = value
                  }
                }
              }
            }
          }
        },
        CommentExit(comment) {
          if (remove && weakMap.get(comment) === cvaSymbol) {
            comment.remove()
          }
        },
        OnceExit(root) {
          result.compoundVariants = [...hashMap.values()]

          process?.({
            base: [...result.base],
            compoundVariants: result.compoundVariants.map((x) => {
              return {
                ...x,
                class: [...x.class],
              }
            }) as CvaParams['compoundVariants'],
            defaultVariants: result.defaultVariants,
            variants: Object.entries(result.variants).reduce<CvaParams['variants']>((acc, [k, v]) => {
              acc[k] = Object.entries(v).reduce<Record<string, string[]>>((bcc, [k1, v1]) => {
                bcc[k1] = [...v1]
                return bcc
              }, {})
              return acc
            }, {}),
            meta: result.meta,
            file: root.source?.input.file,
            // root
          })
        },
      }
    },
  }
}

creator.postcss = true

export default creator
