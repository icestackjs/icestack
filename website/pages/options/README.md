@icestack/ui

# @icestack/ui

## Table of contents

### Type Aliases

- [BaseOptions](README.md#baseoptions)
- [CodegenMode](README.md#codegenmode)
- [CodegenOptions](README.md#codegenoptions)
- [ComponentsOptions](README.md#componentsoptions)
- [ComponentsValue](README.md#componentsvalue)
- [Config](README.md#config)
- [CssInJs](README.md#cssinjs)
- [DeepPartial](README.md#deeppartial)
- [DeepRequired](README.md#deeprequired)
- [GlobalOptions](README.md#globaloptions)
- [ILayer](README.md#ilayer)
- [LoadCodeOptions](README.md#loadcodeoptions)
- [ModeMergeValue](README.md#modemergevalue)
- [TailwindcssPluginOptions](README.md#tailwindcsspluginoptions)

## Type Aliases

### BaseOptions

Ƭ **BaseOptions**\<`T`\>: `Object`

#### Type parameters

| Name | Type |
| :------ | :------ |
| `T` | extends `string` = `string` |

#### Type declaration

| Name | Type |
| :------ | :------ |
| `extraColors` | `Record`\<keyof [`BaseOptions`](README.md#baseoptions)[``"themes"``], `Record`\<`string`, `string`\>\> |
| `extraVars` | `Record`\<keyof [`BaseOptions`](README.md#baseoptions)[``"themes"``], `Record`\<`string`, `string`\>\> |
| `themes` | `Record`\<`T`, \{ `selector`: `string`  }\> |
| `types` | `Record`\<`string`, `Record`\<keyof [`BaseOptions`](README.md#baseoptions)[``"themes"``], `Record`\<`string`, `string`\>\>\> |

#### Defined in

[packages/ui/src/types.ts:18](https://github.com/sonofmagic/icestack/blob/e6652f6/packages/ui/src/types.ts#L18)

___

### CodegenMode

Ƭ **CodegenMode**: ``"styled"`` \| ``"base"`` \| ``"raw"``

#### Defined in

[packages/ui/src/types.ts:67](https://github.com/sonofmagic/icestack/blob/e6652f6/packages/ui/src/types.ts#L67)

___

### CodegenOptions

Ƭ **CodegenOptions**: `Object`

#### Type declaration

| Name | Type |
| :------ | :------ |
| `base` | [`BaseOptions`](README.md#baseoptions) |
| `cache` | `boolean` |
| `components` | [`ComponentsOptions`](README.md#componentsoptions) |
| `dryRun` | `boolean` |
| `global` | [`GlobalOptions`](README.md#globaloptions) |
| `log` | `boolean` |
| `mode` | [`CodegenMode`](README.md#codegenmode) |
| `outdir` | `string` |
| `prefix` | `PrefixerOptions` |
| `presets` | ([`DeepPartial`](README.md#deeppartial)\<[`CodegenOptions`](README.md#codegenoptions)\> \| (`options?`: `any`) => [`DeepPartial`](README.md#deeppartial)\<[`CodegenOptions`](README.md#codegenoptions)\>)[] |
| `tailwindcssConfig` | `TailwindcssConfig` |
| `varPrefix` | `string` |

#### Defined in

[packages/ui/src/types.ts:73](https://github.com/sonofmagic/icestack/blob/e6652f6/packages/ui/src/types.ts#L73)

___

### ComponentsOptions

Ƭ **ComponentsOptions**: `Record`\<`string`, [`ComponentsValue`](README.md#componentsvalue) \| ``false``\>

#### Defined in

[packages/ui/src/types.ts:48](https://github.com/sonofmagic/icestack/blob/e6652f6/packages/ui/src/types.ts#L48)

___

### ComponentsValue

Ƭ **ComponentsValue**: `Object`

#### Type declaration

| Name | Type |
| :------ | :------ |
| `disabled` | `boolean` |
| `extend` | [`ModeMergeValue`](README.md#modemergevalue) |
| `extra` | [`CssInJs`](README.md#cssinjs) |
| `mode` | [`CodegenMode`](README.md#codegenmode) |
| `override` | [`ModeMergeValue`](README.md#modemergevalue) |
| `prefix` | `PrefixerOptions` |
| `schema` | `GetSchemaFn` |
| `selector` | `string` |

#### Defined in

[packages/ui/src/types.ts:37](https://github.com/sonofmagic/icestack/blob/e6652f6/packages/ui/src/types.ts#L37)

___

### Config

Ƭ **Config**: `Partial`\<[`CodegenOptions`](README.md#codegenoptions)\>

#### Defined in

[packages/ui/src/types.ts:102](https://github.com/sonofmagic/icestack/blob/e6652f6/packages/ui/src/types.ts#L102)

___

### CssInJs

Ƭ **CssInJs**: `Record`\<`string`, `any`\>

CSS-in-JS object

#### Defined in

node_modules/.pnpm/@types+postcss-js@4.0.4/node_modules/@types/postcss-js/index.d.ts:5

___

### DeepPartial

Ƭ **DeepPartial**\<`T`\>: \{ [P in keyof T]?: T[P] extends object ? DeepPartial\<T[P]\> : Partial\<T[P]\> }

#### Type parameters

| Name |
| :------ |
| `T` |

#### Defined in

[packages/ui/src/types.ts:118](https://github.com/sonofmagic/icestack/blob/e6652f6/packages/ui/src/types.ts#L118)

___

### DeepRequired

Ƭ **DeepRequired**\<`T`\>: `Required`\<\{ [K in keyof T]: T[K] extends Required\<T[K]\> ? T[K] : DeepRequired\<T[K]\> }\>

#### Type parameters

| Name |
| :------ |
| `T` |

#### Defined in

[packages/ui/src/types.ts:122](https://github.com/sonofmagic/icestack/blob/e6652f6/packages/ui/src/types.ts#L122)

___

### GlobalOptions

Ƭ **GlobalOptions**: `Object`

#### Type declaration

| Name | Type |
| :------ | :------ |
| `atMedia` | \{ `hover`: `boolean`  } |
| `atMedia.hover` | `boolean` |
| `pseudo` | \{ `where`: `boolean`  } |
| `pseudo.where` | `boolean` |
| `selector` | \{ `universal`: `string`  } |
| `selector.universal` | `string` |

#### Defined in

[packages/ui/src/types.ts:50](https://github.com/sonofmagic/icestack/blob/e6652f6/packages/ui/src/types.ts#L50)

___

### ILayer

Ƭ **ILayer**: ``"base"`` \| ``"utilities"`` \| ``"components"``

#### Defined in

[packages/ui/src/types.ts:128](https://github.com/sonofmagic/icestack/blob/e6652f6/packages/ui/src/types.ts#L128)

___

### LoadCodeOptions

Ƭ **LoadCodeOptions**: `Object`

#### Type declaration

| Name | Type |
| :------ | :------ |
| `loadDirectory` | `string` |

#### Defined in

[packages/ui/src/types.ts:69](https://github.com/sonofmagic/icestack/blob/e6652f6/packages/ui/src/types.ts#L69)

___

### ModeMergeValue

Ƭ **ModeMergeValue**: `Object`

#### Type declaration

| Name | Type |
| :------ | :------ |
| `base?` | [`CssInJs`](README.md#cssinjs) |
| `styled?` | [`CssInJs`](README.md#cssinjs) |
| `utils?` | [`CssInJs`](README.md#cssinjs) |

#### Defined in

[packages/ui/src/types.ts:31](https://github.com/sonofmagic/icestack/blob/e6652f6/packages/ui/src/types.ts#L31)

___

### TailwindcssPluginOptions

Ƭ **TailwindcssPluginOptions**: `Object`

#### Type declaration

| Name | Type |
| :------ | :------ |
| `loadDirectory` | `string` |

#### Defined in

[packages/ui/src/types.ts:104](https://github.com/sonofmagic/icestack/blob/e6652f6/packages/ui/src/types.ts#L104)
