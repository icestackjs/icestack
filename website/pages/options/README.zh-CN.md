@icestack/ui

# @icestack/ui

## Table of contents

### Interfaces

- [IBuildScssOptions](interfaces/IBuildScssOptions.md)

### Type Aliases

- [BaseOptions](README.md#baseoptions)
- [CodegenMode](README.md#codegenmode)
- [CodegenOptions](README.md#codegenoptions)
- [ComponentsOptions](README.md#componentsoptions)
- [ComponentsValue](README.md#componentsvalue)
- [Config](README.md#config)
- [DeepPartial](README.md#deeppartial)
- [DeepRequired](README.md#deeprequired)
- [GlobalOptions](README.md#globaloptions)
- [LoadCodeOptions](README.md#loadcodeoptions)
- [ModeMergeValue](README.md#modemergevalue)
- [TailwindcssPluginOptions](README.md#tailwindcsspluginoptions)

## Type Aliases

### BaseOptions

Ƭ **BaseOptions**\<`T`\>: `Object`

#### Type parameters

| Name | Type |
| :------ | :------ |
| `T` | extends `string` = `string` |

#### Type declaration

| Name | Type |
| :------ | :------ |
| `extraColors` | `Record`\<keyof [`BaseOptions`](README.md#baseoptions)[``"themes"``], `Record`\<`string`, `string`\>\> |
| `extraVars` | `Record`\<keyof [`BaseOptions`](README.md#baseoptions)[``"themes"``], `Record`\<`string`, `string`\>\> |
| `themes` | `Record`\<`T`, \{ `selector`: `string`  }\> |
| `types` | `Record`\<`string`, `Record`\<keyof [`BaseOptions`](README.md#baseoptions)[``"themes"``], `Record`\<`string`, `string`\>\>\> |

#### Defined in

[types.ts:17](https://github.com/sonofmagic/icestack/blob/4232246/packages/ui/src/types.ts#L17)

___

### CodegenMode

Ƭ **CodegenMode**: ``"styled"`` \| ``"base"`` \| ``"raw"``

#### Defined in

[types.ts:67](https://github.com/sonofmagic/icestack/blob/4232246/packages/ui/src/types.ts#L67)

___

### CodegenOptions

Ƭ **CodegenOptions**: `Object`

#### Type declaration

| Name | Type |
| :------ | :------ |
| `base` | [`BaseOptions`](README.md#baseoptions) |
| `cache` | `boolean` |
| `components` | [`ComponentsOptions`](README.md#componentsoptions) |
| `dryRun` | `boolean` |
| `global` | [`GlobalOptions`](README.md#globaloptions) |
| `log` | `boolean` |
| `mode` | [`CodegenMode`](README.md#codegenmode) |
| `outdir` | `string` |
| `prefix` | `string` \| `PrefixerOptions` |
| `presets` | [`DeepPartial`](README.md#deeppartial)\<[`CodegenOptions`](README.md#codegenoptions)\>[] |
| `tailwindcssConfig` | `TailwindcssConfig` |
| `varPrefix` | `string` |

#### Defined in

[types.ts:73](https://github.com/sonofmagic/icestack/blob/4232246/packages/ui/src/types.ts#L73)

___

### ComponentsOptions

Ƭ **ComponentsOptions**: `Record`\<typeof `componentsNames`[`number`], [`ComponentsValue`](README.md#componentsvalue)\>

#### Defined in

[types.ts:48](https://github.com/sonofmagic/icestack/blob/4232246/packages/ui/src/types.ts#L48)

___

### ComponentsValue

Ƭ **ComponentsValue**: `Object`

#### Type declaration

| Name | Type |
| :------ | :------ |
| `extend` | [`ModeMergeValue`](README.md#modemergevalue) |
| `extra` | `CssInJs` |
| `mode` | [`CodegenMode`](README.md#codegenmode) |
| `override` | [`ModeMergeValue`](README.md#modemergevalue) |
| `selector` | `string` |

#### Defined in

[types.ts:36](https://github.com/sonofmagic/icestack/blob/4232246/packages/ui/src/types.ts#L36)

___

### Config

Ƭ **Config**: `Partial`\<[`CodegenOptions`](README.md#codegenoptions)\>

#### Defined in

[types.ts:102](https://github.com/sonofmagic/icestack/blob/4232246/packages/ui/src/types.ts#L102)

___

### DeepPartial

Ƭ **DeepPartial**\<`T`\>: \{ [P in keyof T]?: T[P] extends object ? DeepPartial\<T[P]\> : Partial\<T[P]\> }

#### Type parameters

| Name |
| :------ |
| `T` |

#### Defined in

[types.ts:122](https://github.com/sonofmagic/icestack/blob/4232246/packages/ui/src/types.ts#L122)

___

### DeepRequired

Ƭ **DeepRequired**\<`T`\>: `Required`\<\{ [K in keyof T]: T[K] extends Required\<T[K]\> ? T[K] : DeepRequired\<T[K]\> }\>

#### Type parameters

| Name |
| :------ |
| `T` |

#### Defined in

[types.ts:126](https://github.com/sonofmagic/icestack/blob/4232246/packages/ui/src/types.ts#L126)

___

### GlobalOptions

Ƭ **GlobalOptions**: `Object`

#### Type declaration

| Name | Type |
| :------ | :------ |
| `atMedia` | \{ `hover`: `boolean`  } |
| `atMedia.hover` | `boolean` |
| `pseudo` | \{ `where`: `boolean`  } |
| `pseudo.where` | `boolean` |
| `selector` | \{ `universal`: `string`  } |
| `selector.universal` | `string` |

#### Defined in

[types.ts:50](https://github.com/sonofmagic/icestack/blob/4232246/packages/ui/src/types.ts#L50)

___

### LoadCodeOptions

Ƭ **LoadCodeOptions**: `Object`

#### Type declaration

| Name | Type |
| :------ | :------ |
| `loaddir` | `string` |

#### Defined in

[types.ts:69](https://github.com/sonofmagic/icestack/blob/4232246/packages/ui/src/types.ts#L69)

___

### ModeMergeValue

Ƭ **ModeMergeValue**: `Object`

#### Type declaration

| Name | Type |
| :------ | :------ |
| `base?` | `CssInJs` |
| `styled?` | `CssInJs` |
| `utils?` | `CssInJs` |

#### Defined in

[types.ts:30](https://github.com/sonofmagic/icestack/blob/4232246/packages/ui/src/types.ts#L30)

___

### TailwindcssPluginOptions

Ƭ **TailwindcssPluginOptions**: `Object`

#### Type declaration

| Name | Type |
| :------ | :------ |
| `loaddir` | `string` |

#### Defined in

[types.ts:104](https://github.com/sonofmagic/icestack/blob/4232246/packages/ui/src/types.ts#L104)
