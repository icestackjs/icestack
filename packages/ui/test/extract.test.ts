import { getJsObj, extractAll } from '@/extract'
import { getCodegenOptions } from '@/options'
describe('extract', () => {
  it('getJsObj', async () => {
    const obj = await extractAll(getCodegenOptions())
    expect(obj).toBeTruthy()
    expect(obj).toMatchSnapshot()
  })
})
