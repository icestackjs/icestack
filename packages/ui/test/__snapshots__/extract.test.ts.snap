// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`extract > getJsObj base 1`] = `
{
  ".hidden": {
    "display": "none",
  },
  ":root": {
    "--ice-animation-btn": "0.25s",
    "--ice-animation-input": "0.2s",
    "--ice-base-100": "255 255 255",
    "--ice-base-1000": "38 38 38",
    "--ice-base-1100": "31 31 31",
    "--ice-base-1200": "20 20 20",
    "--ice-base-1300": "0 0 0",
    "--ice-base-200": "250 250 250",
    "--ice-base-300": "245 245 245",
    "--ice-base-400": "240 240 240",
    "--ice-base-500": "217 217 217",
    "--ice-base-600": "191 191 191",
    "--ice-base-700": "140 140 140",
    "--ice-base-800": "89 89 89",
    "--ice-base-900": "67 67 67",
    "--ice-base-content": "0 0 0",
    "--ice-border-btn": "1px",
    "--ice-error": "245 34 45",
    "--ice-error-active": "207 19 34",
    "--ice-error-content": "255 255 255",
    "--ice-error-hover": "255 77 79",
    "--ice-info": "22 119 255",
    "--ice-info-active": "9 88 217",
    "--ice-info-content": "255 255 255",
    "--ice-info-hover": "64 150 255",
    "--ice-neutral": "38 38 38",
    "--ice-neutral-active": "31 31 31",
    "--ice-neutral-content": "255 255 255",
    "--ice-neutral-hover": "67 67 67",
    "--ice-primary": "22 119 255",
    "--ice-primary-active": "9 88 217",
    "--ice-primary-content": "255 255 255",
    "--ice-primary-hover": "64 150 255",
    "--ice-rounded-badge": "1.9rem",
    "--ice-rounded-box": "1rem",
    "--ice-rounded-btn": "0.5rem",
    "--ice-success": "82 196 26",
    "--ice-success-active": "56 158 13",
    "--ice-success-content": "255 255 255",
    "--ice-success-hover": "115 209 61",
    "--ice-tab-border": "1px",
    "--ice-tab-radius": "0.5rem",
    "--ice-warning": "250 173 20",
    "--ice-warning-active": "212 136 6",
    "--ice-warning-content": "255 255 255",
    "--ice-warning-hover": "255 197 61",
  },
  "[data-theme=dark]": {
    "--ice-animation-btn": "0.25s",
    "--ice-animation-input": "0.2s",
    "--ice-base-100": "0 0 0",
    "--ice-base-1000": "240 240 240",
    "--ice-base-1100": "245 245 245",
    "--ice-base-1200": "250 250 250",
    "--ice-base-1300": "255 255 255",
    "--ice-base-200": "20 20 20",
    "--ice-base-300": "31 31 31",
    "--ice-base-400": "38 38 38",
    "--ice-base-500": "67 67 67",
    "--ice-base-600": "89 89 89",
    "--ice-base-700": "140 140 140",
    "--ice-base-800": "191 191 191",
    "--ice-base-900": "217 217 217",
    "--ice-base-content": "255 255 255",
    "--ice-border-btn": "1px",
    "--ice-error": "211 32 41",
    "--ice-error-active": "232 71 73",
    "--ice-error-hover": "166 29 36",
    "--ice-info": "22 104 220",
    "--ice-info-active": "60 137 232",
    "--ice-info-content": "255 255 255",
    "--ice-info-hover": "21 84 173",
    "--ice-neutral": "38 38 38",
    "--ice-neutral-active": "31 31 31",
    "--ice-neutral-content": "255 255 255",
    "--ice-neutral-hover": "67 67 67",
    "--ice-primary": "22 104 220",
    "--ice-primary-active": "60 137 232",
    "--ice-primary-content": "255 255 255",
    "--ice-primary-hover": "21 84 173",
    "--ice-rounded-badge": "1.9rem",
    "--ice-rounded-box": "1rem",
    "--ice-rounded-btn": "0.5rem",
    "--ice-success": "73 170 25",
    "--ice-success-active": "106 190 57",
    "--ice-success-content": "255 255 255",
    "--ice-success-hover": "60 134 24",
    "--ice-tab-border": "1px",
    "--ice-tab-radius": "0.5rem",
    "--ice-warning": "216 150 20",
    "--ice-warning-active": "232 179 57",
    "--ice-warning-content": "255 255 255",
    "--ice-warning-hover": "170 119 20",
  },
}
`;
